[GTK4] Make favicon and snapshot API use GdkTexture instead of cairo surfaces

carlosgcampos opened the pull request:
#### d15c97f78b0303aabe2cecbb9e1e7f67ab69a028
<pre>
[GTK4] Make favicon and snapshot API use GdkTexture instead of cairo surfaces
<a href="https://bugs.webkit.org/show_bug.cgi?id=226333">https://bugs.webkit.org/show_bug.cgi?id=226333</a>

Reviewed by Adrian Perez de Castro.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/platform/graphics/gtk/GdkCairoUtilities.cpp:
(WebCore::cairoSurfaceToGdkTexture):
* Source/WebCore/platform/graphics/gtk/GdkCairoUtilities.h:
* Source/WebCore/platform/graphics/gtk/ImageGtk.cpp:
(WebCore::BitmapImage::gdkTexture):
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
(webkitFaviconDatabaseSetIconForPageURL):
(webkit_favicon_database_get_favicon_finish):
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewUpdateFavicon):
(webkitWebViewCancelFaviconRequest):
(gotFaviconCallback):
(webkitWebViewGetProperty):
(webkit_web_view_class_init):
(webkit_web_view_get_favicon):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Tools/MiniBrowser/gtk/BrowserWindow.c:
(faviconChanged):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitFaviconDatabase.cpp:
(testFaviconDatabaseGetFavicon):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitFindController.cpp:
(testFindControllerHide):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
(testWebViewSnapshot):
* Tools/TestWebKitAPI/glib/WebKitGLib/WebViewTest.cpp:
(WebViewTest::~WebViewTest):
(onSnapshotReady): Deleted.
(WebViewTest::getSnapshotAndWaitUntilReady): Deleted.
* Tools/TestWebKitAPI/glib/WebKitGLib/WebViewTest.h:

Canonical link: <a href="https://commits.webkit.org/259997@main">https://commits.webkit.org/259997@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ca9517266cad4f181001645e21d056d5e40092b1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106536 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15552 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39345 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115721 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115366 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110445 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17047 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6780 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98728 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112304 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95921 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40508 "Passed tests") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27565 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82231 "Found 1 new API test failure: /WebKitGTK/TestDownloads:/webkit/Downloads/local-file-error (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8781 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28917 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9324 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5636 "Found 1 new test failure: fast/text/glyph-display-lists/glyph-display-list-scaled-unshared.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14940 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48465 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10862 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3718 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

aperezdc reviewed (approved):
When I read the PR title, I thought this would be gnarlier. The changes looks reasonably contained 👌🏼

mcatanzaro commented on Source/WebCore/platform/graphics/gtk/GdkCairoUtilities.cpp:
```diff
@@ -146,4 +146,19 @@ GdkPixbuf* cairoSurfaceToGdkPixbuf(cairo_surface_t* surface)
 }
 
+#if USE(GTK4)
+GdkTexture* cairoSurfaceToGdkTexture(cairo_surface_t* surface)
```
```suggestion
GRefPtr<GdkTexture> cairoSurfaceToGdkTexture(cairo_surface_t* surface)
```

carlosgcampos replied:
Since the same applies to cairoSurfaceToGdkPixbuf, I'll do this for both in a follow up.

mcatanzaro commented on Source/WebCore/platform/graphics/gtk/GdkCairoUtilities.cpp:
```diff
@@ -146,4 +146,19 @@ GdkPixbuf* cairoSurfaceToGdkPixbuf(cairo_surface_t* surface)
+    GRefPtr<GBytes> bytes = adoptGRef(g_bytes_new_with_free_func(data, height * stride, [](gpointer data) {
+        cairo_surface_destroy(static_cast<cairo_surface_t*>(data));
+    }, cairo_surface_reference(surface)));
+    return gdk_memory_texture_new(width, height, GDK_MEMORY_DEFAULT, bytes.get(), stride);
```
```suggestion
    return adoptGRef(gdk_memory_texture_new(width, height, GDK_MEMORY_DEFAULT, bytes.get(), stride));
```

mcatanzaro commented on Source/WebCore/platform/graphics/gtk/GdkCairoUtilities.h:
```diff
@@ -28,5 +28,7 @@
 GdkPixbuf* cairoSurfaceToGdkPixbuf(cairo_surface_t*);
-
+#if USE(GTK4)
+GdkTexture* cairoSurfaceToGdkTexture(cairo_surface_t*);
```
```suggestion
GRefPtr<GdkTexture> cairoSurfaceToGdkTexture(cairo_surface_t*);
```

mcatanzaro replied:
```suggestion
GRefPtr<GdkTexture> cairoSurfaceToGdkTexture(cairo_surface_t*);
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
```diff
@@ -238,21 +233,24 @@ void webkit_favicon_database_get_favicon(WebKitFaviconDatabase* database, const
-}
+#if USE(GTK4)
+    auto icon = adoptRef(static_cast<cairo_surface_t*>(g_task_propagate_pointer(G_TASK(result), error)));
+    return icon ? cairoSurfaceToGdkTexture(icon.get()) : nullptr;
```
```suggestion
    return icon ? cairoSurfaceToGdkTexture(icon.get()).leakRef() : nullptr;
```

Yes really... even though the smart pointer kind of gets in the way here, it's important to use them to transfer ownership wherever possible, especially when the function is defined in an entirely different file.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1210,10 +1228,18 @@ static void webkit_web_view_class_init(WebKitWebViewClass* webViewClass)
      * See webkit_web_view_get_favicon() for more details.
      */
     sObjProperties[PROP_FAVICON] =
+#if USE(GTK4)
+        g_param_spec_object(
+            "favicon",
+            nullptr, nullptr,
+            GDK_TYPE_TEXTURE,
+            WEBKIT_PARAM_READABLE);
+#else
         g_param_spec_pointer(
             "favicon",
             nullptr, nullptr,
             WEBKIT_PARAM_READABLE);
+#endif
```
I wonder if gir-scanner will be able to handle this. We know it cannot handle conditional compilation for signals, so I suspect it will trip over properties as well. Probably should split out into the WebKitWebViewGtk3.cpp and WebKitWebViewGtk4.cpp?

carlosgcampos replied:
I think the problem is conditional compilation of the documentation comment, not the definition itself. In this case the doc is the same.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -4851,14 +4881,23 @@ void webkit_web_view_get_snapshot(WebKitWebView* webView, WebKitSnapshotRegion r
 
+#if USE(GTK4)
+    auto image = adoptRef(static_cast<cairo_surface_t*>(g_task_propagate_pointer(G_TASK(result), error)));
+    return image ? cairoSurfaceToGdkTexture(image.get()) : nullptr;
```
```suggestion
    return image ? cairoSurfaceToGdkTexture(image.get()).leakRef() : nullptr;
```

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitFindController.cpp:
```diff
@@ -60,9 +60,33 @@ class FindControllerTest: public WebViewTest {
     }
 
+#if PLATFORM(GTK)
+    cairo_surface_t* takeSnapshotAndWaitUntilReady()
```
```suggestion
    GRefPtr<cairo_surface_t> takeSnapshotAndWaitUntilReady()
```

Need to add a new GRefPtr specialization.

carlosgcampos replied:
No, GRefPtr is for glib types, cairo surface is not. We already have a RefPtr for cairo, but it's in WebCore, because WTF doesn't depend on cairo. The api tests only use the public API and WTF, but not WebCore API. That's the reason why for cairo types we manually ref/unref. Leaks in api tests that are no related to public API are not problematic, so I don't think it's worth the effort.

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitFindController.cpp:
```diff
@@ -60,9 +60,33 @@ class FindControllerTest: public WebViewTest {
     bool m_textFound;
     unsigned m_matchCount;
+#if PLATFORM(GTK)
+    cairo_surface_t* m_surface { nullptr };
```
```suggestion
    GRefPtr<cairo_surface_t> m_surface;
```

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1087,52 +1135,106 @@ class SnapshotWebViewTest: public WebViewTest {
+        return surface;
+    }
+#else
+    static cairo_surface_t* snapshotToSurface(cairo_surface_t* snapshot)
```
```suggestion
    static GRefPtr<cairo_surface_t> snapshotToSurface(cairo_surface_t* snapshot)
```

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1087,52 +1135,106 @@ class SnapshotWebViewTest: public WebViewTest {
+#if USE(GTK4)
+    GRefPtr<GdkTexture> m_snapshot;
+#else
+    cairo_surface_t* m_snapshot { nullptr };
```
```suggestion
    GRefPtr<cairo_surface_t> m_snapshot;
```

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1065,20 +1065,68 @@ class SnapshotWebViewTest: public WebViewTest {
+#if USE(GTK4)
+    GdkTexture* getSnapshotAndWaitUntilReady(WebKitSnapshotRegion region, WebKitSnapshotOptions options)
+#else
+    cairo_surface_t* getSnapshotAndWaitUntilReady(WebKitSnapshotRegion region, WebKitSnapshotOptions options)
```
```suggestion
    GRefPtr<cairo_surface_t> getSnapshotAndWaitUntilReady(WebKitSnapshotRegion region, WebKitSnapshotOptions options)
```

mcatanzaro reviewed (commented):
So the change is good, but I think you should add a GRefPtr specialization for cairo_surface_t instead of transferring ownership via raw pointer all over the place, especially in the tests. You have a lot of opportunities to get rid of manual cairo_surface_destroy().

Same for GdkTexture, except that is a GObject so no extra specialization required there.

carlosgcampos committed:
[GTK4] Make favicon and snapshot API use GdkTexture instead of cairo surfaces
https://bugs.webkit.org/show_bug.cgi?id=226333

Reviewed by Adrian Perez de Castro.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/platform/graphics/gtk/GdkCairoUtilities.cpp:
(WebCore::cairoSurfaceToGdkTexture):
* Source/WebCore/platform/graphics/gtk/GdkCairoUtilities.h:
* Source/WebCore/platform/graphics/gtk/ImageGtk.cpp:
(WebCore::BitmapImage::gdkTexture):
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
(webkitFaviconDatabaseSetIconForPageURL):
(webkit_favicon_database_get_favicon_finish):
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewUpdateFavicon):
(webkitWebViewCancelFaviconRequest):
(gotFaviconCallback):
(webkitWebViewGetProperty):
(webkit_web_view_class_init):
(webkit_web_view_get_favicon):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Tools/MiniBrowser/gtk/BrowserWindow.c:
(faviconChanged):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitFaviconDatabase.cpp:
(testFaviconDatabaseGetFavicon):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitFindController.cpp:
(testFindControllerHide):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
(testWebViewSnapshot):
* Tools/TestWebKitAPI/glib/WebKitGLib/WebViewTest.cpp:
(WebViewTest::~WebViewTest):
(onSnapshotReady): Deleted.
(WebViewTest::getSnapshotAndWaitUntilReady): Deleted.
* Tools/TestWebKitAPI/glib/WebKitGLib/WebViewTest.h:

Canonical link: https://commits.webkit.org/259997@main
